cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/DynamicPartitioner.java

```diff
@@ -27,6 +28,25 @@
  */
 public abstract class DynamicPartitioner<K, V> {
 
+  /**
+   * Defines options for when writing to a partition.
+   */
+  public enum PartitionWriteOption {
+    /**
+     * Will only allow writing to a new partition. If an attempt is made to write to a previously-existing partition,
+     * a {@link DataSetException} will be thrown.
+     */
+    NEW_ONLY,
+    /**
+     * Allows appending to an existing partition. If the partition does not already exist, it will be created.
+     */
+    ALLOW_APPEND,
```

Comment:
I think it's better to keep it like file create attribute that can be combined. E.g. `CREATE`, `APPEND`, `OVERWRITE`. When using it, it can be `(CREATE)` for new only or `(CREATE, APPEND)` for create if not exist, otherwise append, etc.